src/titiler/xarray/titiler/xarray/extensions.py

```diff
@@ -36,3 +37,55 @@ def variables(
             """return available variables."""
             with self.dataset_opener(src_path, **io_params.as_dict()) as ds:
                 return list(ds.data_vars)  # type: ignore
+
+
+@define
+class DatasetMetadataExtension(FactoryExtension):
+    """Add dataset metadata endpoints to a Xarray TilerFactory."""
+
+    io_dependency: Type[DefaultDependency] = XarrayIOParams
+    dataset_opener: Callable[..., xarray.Dataset] = xarray_open_dataset
+
+    def register(self, factory: TilerFactory):
+        """Register endpoint to the tiler factory."""
+
+        @factory.router.get(
+            "/keys",
+            response_model=List[str],
+            responses={200: {"description": "Return Xarray Dataset variables."}},
+        )
+        def dataset_variables(
+            src_path=Depends(factory.path_dependency),
+            io_params=Depends(self.io_dependency),
+        ):
+            """return available variables."""
+            with self.dataset_opener(src_path, **io_params.as_dict()) as ds:
+                return list(ds.data_vars)
+
+        @factory.router.get(
+            "/dict",
+            response_model=Dict[str, Any],
+            responses={200: {"description": "Return Xarray Dataset metadata as Dict."}},
+        )
+        def dataset_to_dict(
+            src_path=Depends(factory.path_dependency),
+            io_params=Depends(self.io_dependency),
+        ):
+            """return dataset as dict."""
+            with self.dataset_opener(src_path, **io_params.as_dict()) as ds:
+                return ds.to_dict(data=False)
+
+        @factory.router.get(
+            "/html",
+            response_class=HTMLResponse,
+            responses={
+                200: {"description": "Return Xarray Dataset metadata as HTML Document."}
+            },
+        )
```

Comment:
Could we have a single endpoint (e.g. `/info`) then have a format parameter that determines the response type. Maybe like we do in the eoapi-devseed STAC endpoints that return html for browser sessions, json otherwise.

Reply:
Yeah I wondered about that, it's appealing. 

We can't have an `/info` endpoint because we already include an /info endpoint with the regular factory https://github.com/developmentseed/titiler/blob/57b44d6f2fc32153705550643eb61913f524935f/src/titiler/xarray/titiler/xarray/factory.py#L62-L131

I'm super open about the endpoint names 

we could also imagine to have a `/variables/{variables}`  prefix to all the endpoint that needs a `variable` input e.g : `/variables/lst/info?url=https://gws-access.jasmin.ac.uk/public/leicester/SENTINEL3A_SLSTR_L3C_0.01/4.00/2024/10/16/EOCIS_-LST-L3C-LST-SLSTRA-0.01deg_1DAILY_DAY-20241016000000-fv4.00.nc`

Reply:
@hrodmn I've updated this PR:
- renamed `/dict` and `/html` to `/info` and apply dynamic content-type (like int stac-html) 
- added `/dataset` prefix

Reply:
or maybe we could follow xpublish 
- `/dataset/`  (HTML, `ds._repr_html_()`)
- `/dataset/dict` (JSON, `ds.to_dict(data=False)`)
- `/dataset/keys (JSON, `list(ds.data_vars)`)
- `/dataset/info` (JSON, 🤷)

Reply:
+1 on aligning with xpublish on endpoint names

Reply:
I love the smart response type logic! Following xpublish seems like the a good idea since these endpoints mirror the same functionality as xpublish, but is xpublish mature enough to rely on for this purpose yet? Is that package following any kind of spec? I realize we aren't following a spec in these titiler endpoints either.

Reply:
Yeah, to me it's the same as the Egg or Chicken question, who have to follow who 😅 

I'm happy with both solution, but I just want to make sure it makes sense from the user POV

Reply:
I think I'm going to go with the `xpublish` way (mostly because it adds an open door to add `/dataset/info`)